packages/utilities/crypto/lib/utils.ts

```diff
@@ -28,16 +28,145 @@ export const concatBuffers = (bufs: Uint8Array[]): Uint8Array => {
 	return bytes;
 };
 
-export const isBufferEqual = (a: Uint8Array, b: Uint8Array): boolean => {
+// -- Cryptographic commons --
+
+export const getSignatureS = (sig: Uint8Array): bigint => {
+	// Low-level signature operation; you likely want to use `normalizeSignature` and `isSignatureNormalized` instead.
+
+	// Reference: [1] RFC 7518 JSON Web Algorithms (JWA), § 3.4. Digital Signature with ECDSA -- https://datatracker.ietf.org/doc/html/rfc7518#section-3.4
+	// Trivia:    This format is also known as IEEE P1363, however the standard is paid (and inactive).
+	// We want the upper-half of the sig, which contains s according to [1]
+	let len: number;
+	let s = 0n;
+	for (let i = 1; i <= (len = sig.length) >> 1; i++) {
+		// s is encoded as big-endian; c.f. [1] Point 2.
+		s |= BigInt(sig[len - i]) << BigInt(8 * (i - 1));
+	}
+
+	return s;
+};
+
+export const setSignatureS = (sig: Uint8Array, s: bigint) => {
+	// Low-level signature operation; you likely want to use `normalizeSignature` and `isSignatureNormalized` instead.
+
+	// Reference: see getSignatureS
+	// We target the upper-half of the sig, which contains s according to [1]
 	let len: number;
+	for (let i = 1; i <= (len = sig.length) >> 1; i++) {
+		// s is encoded as big-endian; c.f. [1] Point 2.
+		sig[len - i] = Number((s >> BigInt(8 * (i - 1))) & 0xffn);
+	}
+};
+
+export const isSignatureNormalized = (sig: Uint8Array, curveOrder: bigint): boolean => {
+	// Reference: Bitcoin BIP 0146 -- https://github.com/bitcoin/bips/blob/665712c/bip-0146.mediawiki#low_s
+	// Note: Upper bound is inclusive.
+	return getSignatureS(sig) <= curveOrder >> 1n;
+};
+
+export const normalizeSignature = (sig: Uint8Array, curveOrder: bigint) => {
+	// Reference: [1] Bitcoin BIP 0146 -- https://github.com/bitcoin/bips/blob/665712c/bip-0146.mediawiki#low_s
+	//            [2] SEC 1, ver. 2.0, §4.1.3 Signing Operation -- https://www.secg.org/sec1-v2.pdf
+
+	// 1. Retrieve the s value of the (r, s) pair.
+	const s = getSignatureS(sig);
+
+	// 2. Check if s is greater than half of the curve order.
+	//    IF it is (sig is high-S as defined in [1]), then normalize it to low-S as defined in [1].
+	// Note: Upper bound is inclusive.
+	if (s > curveOrder >> 1n) {
+		// [2] "The signer may replace (r, s) with (r, −s mod n), because this is an equivalent signature."
+		// Stmt: Given `s > N / 2`, s is valid; then `-s mod N` is trivially equivalent to `N - s`.
+		setSignatureS(sig, curveOrder - s);
+	}
+
+	// Return the signature to save the need for an extra reference when calling this function.
+	return sig;
+};
+
+export const mutableCompressPoint = (coords: Uint8Array): Uint8Array => {
+	// Reference: [1] SEC 1, ver. 2.0, §2.3.3 Elliptic-Curve-Point-to-Octet-String Conversion -- https://www.secg.org/sec1-v2.pdf
+	// This function MUTATES the input buffer and returns a new view of the original buffer.
```

Comment:
Is there any particular advantage to this mutation? It looks fine as-is but I think it'd be harder to misuse if there was no mutation (i.e. creating a brand new buffer to hold the result)

Reply:
Slightly less code and no memory allocation (besides the new view), nothing more. It's indeed not super pretty to mutate input variables, but in this case the transformation is not destructive, and the original buffer could theoretically be used as-is, just with 32 bytes of extra data.

That being said, looking at it further it'd be trivial to make this function pure by using slice instead, which returns a copy of the region (contrary to subarray which returns a narrow view of the original buffer) 🤔

Reply:
Resolved in 841e3c8ade57133a0e10f30555168b87d8e35961

Reply:
would use `slice()`. don't think the memory allocation reduction is worth the potential side effects

Reply:
thx john github the resolution wasn't there when it posted 😑